src/libraries/System.Private.CoreLib/src/System/Buffers/Text/Utf8Parser/Utf8Parser.Number.cs

```diff
@@ -61,7 +61,7 @@ private static bool TryParseNumber(ReadOnlySpan<byte> source, ref Number.NumberB
             int maxDigitCount = digits.Length - 1;
 
             // Throw away any leading zeroes
-            while (srcIndex != source.Length)
+            while (srcIndex < source.Length)
```

Comment:
Do these need `uint`-casts to eliminate the bound checks?

Reply:
In this case casts appear to regress bounds check elimination.

Perhaps this is unnecessary because `srcIndex` is a local variable that is zero-initialized?

https://github.com/dotnet/runtime/blob/ce18ccfcc350b2ddd258b2429d96c729e949a686/src/libraries/System.Private.CoreLib/src/System/Buffers/Text/Utf8Parser/Utf8Parser.Number.cs#L34

Reply:
We already do checks on `source.Length` before the loop. Adding uint casts seems to cause JIT to "forget" about these.

Reply:
I still don't understand what you mean.  Can you share a sharplab.io example?

Reply:
https://sharplab.io/#v2:EYLgxg9gTgpgtADwGwBYA0AXEBDAzgWwB8ABAJgEYBYAKGIGYACMhgYQYG8aHuGAHKAJYA3bBhhNySBsAgQANgwCy5ABQAlGNgAmAeQB2cgJ4BlXtj0AeYIbEA+BrggBXKGBhoGzjAwF7v1sVwWCD1cJ3wYLQBKLh5Oah5EnwAzBhVHFzcAOgAZGD0AcwwACwYAXjKGAAYYhKTuePr6gJggkLCIrXLqgG5Ypp5iAHYGZOw5XBg+uvqAX36k329cVwBJPS0YBG6q6YXEgHoD1nbw8RLxOU0tXwKHAQK9FIZzQyz9nhaGMG6M1xgANorMDrTYIAC60wGAlSKh+FTSLSiAHIANTI2oDRoDHjA0FbVGoqE47gwtJ4jZbcqVP7ZPKFEqYkkcD4klptUJnLqVXasnHDUbjSbEknzGY4+EOZz/IFrSkQkVJMUDI4MAAqxSgEAA7i9tdhDC89IartpbgwAF4wLWtPna4oCK5pFROJZRClghgWNKuvzu6V0/JFYpM+rYkmS2mAj1bSF8xJkuEMACElWRVQx8aawFg2AA1oqBjGEITCzwxXz2cFOZ1fnKwWXuAKMFAnFMFsqePxhKJxMRJNJZApFFV1Nd9EZTOYrDYYPYox4vD4/NJZxyOpFQyzxQnYVHckGStTqlvEuGcVXTrWeY2kgKxhN2zvy3ylg561SbzQ+arqxuGBcDCmjchT3I8zyvO8z7cF8kYBtGH4Kq+sLwpUKhImimbQduzLviC8qllmpJ7ohx77vSwank0564ZeNaRDst4DPeQpPrhnYRr88GyvhYJxthHFJKqGparq2D6oarxAdc5pWjauB2g6TrpKR3rkYeIZ8jRTRwZkCG8bGTE8ImPypgw6ZYbhiQ5poBZEXh+IlkSfIVthdEbtyDnyrezatmxL7ULMQA=

Reply:
I see.  This appears to be the issue where the JIT won't eliminate bounds checks based on such a condition in a while loop (@EgorBo, I know we've discussed this before but I can't find the issue for it now), but the while loop without the casts is structured in the same pattern that it already recognizes as part of typical for loop looping.

Reply:
Elide bounds checks for partial array/span iteration #31638?

Reply:
It's worth noting that SharpLab is .NET 7

We've made ~6-8 months of additional improvements on top of that already, including a PR that adds JIT support for recognizing that `Span/ROSpan<T>` is considered "never negative" (in the same way it already had support for `T[]`)

So it is very likely worth checking codegen using a .NET 8 preview or better nightly build. You can then use https://marketplace.visualstudio.com/items?itemName=EgorBogatov.Disasmo in VS to easily get the disassembly.

Reply:
> So it is very likely worth checking codegen using a .NET 8 preview or better nightly build. 

I already checked.  On a very recent build, this:
```C#
private static bool M0(ReadOnlySpan<byte> source, out int bytesConsumed)
{
    int srcIndex = 0;

    while (srcIndex < source.Length)
    {
        if (source[srcIndex] != '0')
            break;
        srcIndex++;
    }

    bytesConsumed = srcIndex;
    return true;
}
```
produces:
```asm
; Method C:M0(System.ReadOnlySpan`1[ubyte],byref):bool
G_M000_IG01:                ;; offset=0000H

G_M000_IG02:                ;; offset=0000H
       488B01               mov      rax, bword ptr [rcx]
       8B4908               mov      ecx, dword ptr [rcx+08H]
       4533C0               xor      r8d, r8d
       85C9                 test     ecx, ecx
       7E12                 jle      SHORT G_M000_IG04
                            align    [0 bytes for IG03]

G_M000_IG03:                ;; offset=000DH
       458BC8               mov      r9d, r8d
       42803C0830           cmp      byte  ptr [rax+r9], 48
       7508                 jne      SHORT G_M000_IG04
       41FFC0               inc      r8d
       443BC1               cmp      r8d, ecx
       7CEE                 jl       SHORT G_M000_IG03

G_M000_IG04:                ;; offset=001FH
       448902               mov      dword ptr [rdx], r8d
       B801000000           mov      eax, 1

G_M000_IG05:                ;; offset=0027H
       C3                   ret      
; Total bytes of code: 40
```
while just changing `srcIndex < source.Length` to `(uint)srcIndex < (uint)source.Length` yields:
```asm
; Method C:M1(System.ReadOnlySpan`1[ubyte],byref):bool
G_M000_IG01:                ;; offset=0000H
       4883EC28             sub      rsp, 40

G_M000_IG02:                ;; offset=0004H
       488B01               mov      rax, bword ptr [rcx]
       8B4908               mov      ecx, dword ptr [rcx+08H]
       4533C0               xor      r8d, r8d
       85C9                 test     ecx, ecx
       7626                 jbe      SHORT G_M000_IG04
       0F1F80000000000F1F840000000000 align    [15 bytes for IG03]

G_M000_IG03:                ;; offset=0020H
       443BC1               cmp      r8d, ecx
       731F                 jae      SHORT G_M000_IG06
       458BC8               mov      r9d, r8d
       42803C0830           cmp      byte  ptr [rax+r9], 48
       7508                 jne      SHORT G_M000_IG04
       41FFC0               inc      r8d
       443BC1               cmp      r8d, ecx
       72E9                 jb       SHORT G_M000_IG03

G_M000_IG04:                ;; offset=0037H
       448902               mov      dword ptr [rdx], r8d
       B801000000           mov      eax, 1

G_M000_IG05:                ;; offset=003FH
       4883C428             add      rsp, 40
       C3                   ret      

G_M000_IG06:                ;; offset=0044H
       E8D7E9C35F           call     CORINFO_HELP_RNGCHKFAIL
       CC                   int3     
; Total bytes of code: 74
```

Reply:
Our bound check elimination is done in two different phase:
1) Loop Clonning
2) RangeCheck

2nd is quite powerful and is able to handle many complicated cases but not loops 
1st is pretty fragile, any change to `for (int i = 0; i < arr.Length; i++)` brakes it 😞 cc @dotnet/jit-contrib

Reply:
Thanks for checking! I'd guess the latter is due to us not having "range propagation" and so extra cast is likely causing us to lose track that the value should be "never negative".

Reply:
Filed https://github.com/dotnet/runtime/issues/82946, once loop clonning for spans is implemented this issue will be fixed implicitly. E.g. do the same for arrays and it will be handled just fine

Reply:
> It's worth noting that SharpLab is .NET 7

@tannergooding For future reference, does the codegen from Compiler Explorer have the same optimizations as a local disassembly using .NET 8 nightly build.

Currently, .e.g. https://csharp.godbolt.org/z/cKhTfjvqd, seems to be using:
[`// crossgen2 8.0.0-preview.3.23152.99+8d2ddc45e8f6f2a726c2b91c1ba2b9f7208b1f3b`](8d2ddc45e8f6f2a726c2b91c1ba2b9f7208b1f3b)

Reply:
@xtqqczze for godbolt you need to pass `-O` or better `--Ot` argument to compiler to enable optimizations. godbolt uses crossgen (R2R) which is, on average, worse than what JIT emits (sharplab). Although, even sharplab's codegen is a bit synthetic since in the real world our JIT is tiered and Tier1 codegen might be different from what you see in sharplab.

Reply:
> for godbolt you need to pass -O or better --Ot

@EgorBo Not seeing any differeences in codegen for this example: https://csharp.godbolt.org/z/cKhTfjvqd, is this expected?

Reply:
Yes. However, its showing the crossgen code and so isn't strictly the same as what sharplab shows or what a user will see for Tier 1 codegen.

You can customize the instruction set targeted and I'd recommend using `--instruction-set x86-x64-v3` as that targets `AVX2, BMI1, BMI2, LZCNT, MOVBE, and FMA`

You can customize the OS: `--targetos windows` (defaults to `linux`)
and the architecture: `--targetarch arm64` (defaults to `x64`)

and in general pass in other options supported by crossgen

Reply:
@tannergooding Great help with your tips, I knew about the parameters but couldn't find a useful reference with a search engine.